### ISSUE TITLE
JIT: Inefficient Register Allocation without local var copy

### ISSUE THREAD

**Issue by Tornhoof** [dotnet/runtime#39349]
**tl;dr:**
The JIT on x64 does not assign a register to a byref variable leading to inefficent machine code in a fairly tight loop, this results in a performance which is twice as slow as a version with a local var copy.

The main difference in the generated asm output is: `` add dword ptr [rax], 8`` instead of ``add r11d, 8`` 
[Link to SharpLab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABABgAJiBGAOgCUBXAOwwEt8YaBhCfAB1YAbGFADKIgG6swMXAG4AsAChyq4gGZKAJnJ0YfKBGWrVAb2MnLB1hOwYY5SE1wZyDQRCYBzcgGkKALzkpAhcABzqAIIAnFQAolSKKpYm1rb2jp4ubh7eflTkQSFhqGEAIgBClUkpqVA2dg5O2e6ePr46RQgA7FVxXKTqFTW15GmNmc6urXm+ml1aAGIArGHdpJHLSRa14xmw2AAmnoIAnuTAp/YA2gC65AD6ZXbYheRMMADuF1cw11SkLQoW7bZK1DSUFDkADqrAwAAsADIQMDYQQ8PinAAUAEodilzGDRqoZj4AGasKAuUQYCZFEbEkm5Hy4GBOQ40unBBmM0nkBGUjm0jL0/HEvlkiAMKAIzki7li0asFjkCBksms1yionE5WuSDMLXkAFAnnE2xQcjKvgMI1PF5m0awMnkC3kFzYGVvZ1Wpg2jDXUggxW1CruADWyIgfAAcl8sT6KVSMHKYGhyD7WezU+mfQKoELGrmYC7JdLZcK0xmS+7aTLiy61RqYBg0CHGQaWDjHbVFsq0awAF4wAASeHhCZrSeplYb7rZnkL9jnHvr1cb6s16c7GG77YAvsp2xDiFDYQipRhkaj0dHsXidSZCYyTBLKTOuaQeyk+VnF6m3i/dtLD5fMlwcbUXyZNpyDLGV4QAyCoL1VVNxbQDv0sFCdzeE0UEwkw3WtW03ntWkCNUH03VXI0fWIgMgwo8gw0ESMIGjAB5QRDknUt3xTWd13nbNBLzeFBRzIS4IrIshJoucmy3YDRh3PdH17ftBCHUdx142D+MkzMFyYcCVzrVshMUlttylLtv0PJR22uABZFt4QgQ4AEl+EELFXIvLyfI4vg2CyGhIi8LxYFwXAbBgTymC0phlS8HFbnbPYHGoJBIWYiMoz4LieJ9N9k0Mmtf2M0yhNA8SCxzZTKIq5lYKleDypdS4MmAOcUKsizsNs3d22fKDPnE4RyCxfryAAHiCHC4HIdQtAfKDyFG9bVGnASJgAaiCLg8AwWbSQAPj0nrUObXdyAAKj8ID1MZGaDvIMI5HIAB6L7yCOQ5yEOT5oABkLLWuKBsAQW50zCRqUh2xDdAgSs6FYLx4QwLFEcErRohxcg9v5OrwKYyw/xMgC3qOlxTuZC6fSu/qCYe3xEnhyxXqCD6fr+w4AaBkGxgwcHIeh2GOZMCnwLeOgUcaNGMax6XJLxgmiekhDKzJkwwKpw7jrptoGZrJm0Nu1mtB11QufeuRef+wHgYLYXRahmH3sl1Q9crWX5fsRXMaxH3ZLVwn9LK7Wvda8stf2g3afOy702Z+6/HUa3rs1Qnuft37HcFl2wfICH3Yl57iU1pG5dR9Gg6r3H8fDlWo4rkwHOJBynP89zAr4Xye487z++C0LnHCyLotiiR4sS5UUrSjL6nSLKqBy088tYgq40+PTSo/ZcapalvZJKlqQ+XeGz5ghvT9N34Ll6lUKSYNEOPN9NBsNNbRk2xk3RmkEF+b9zZMXGkIBw01zZzQWkNcgS0Vo/xfH/daON47kBpidJOjMU7mxZo9TOtsPrfXzvzDMVAqCHHLltEwaD5Q1wVnXLGdCqxhyJhfGAmcT4ZGpobbBptcE3XwWzQh0C3rEIdmQqAFCqGezbrUbhEFka1yVliRR6Y2Ex3aq3Gh3sSb6wwXw+mycs4tmEVbaORDPqSIBtIyh1DdEcL9iooOHCNFNw1gZHRujb48ITlg4xODTEW3TqIm6Oc7YkL5rYmRDiaG+KUQwgOTDsZtRkofTRiimJd3kfpV+gh37hKCP1eyR5nqZUoGvXKfZ8naTHLgCc188gsLnJVESd9OoPyunRFU/U2xEi/l2Ea8M3QwBMm8V6U0cIADJlpUDUvDVgLosRjIBktGaZ0ghryQYyFBUEWERMwUbbwJtOqCM1OYsJ2dxGZwOUEJJMBA7MK8bJdQ6hhEZ2joow5RjjYmNTpbK56EblfKqtXf2jyUnqOWu8tOcxbkvIyAAPXuRCp52NEUOEts3MFgkUDLGEezXJ5NcUTBRcoxhqjvms00J4yOsl8XCKeutHJ60llTVWfA4J5BNnvR2cSPZL4Dm8MToEgRwTLmWLEbnaOdyKXJNUS0mFHyEX0uRailxzy1VYr8DoIm0KtDdEJdkspbLlmcvWdA3lKB+W/1Be0vxhjE56lOY/CVcKLHEpMLbfC9r/y+w1ZSoO0K3kqr9ZTX25KHnoupencOSqUBICZSaxyuT2UrPGZa8JvLVrw0FYyYV/jToNOgBgV1ZshEeqBa4N6nrdERwPoktFKSlWhrhZ8r121MXkCjc2xV3bsX6tJRk1acKiUstNchc1mbuW8vmXm8NMs3oAFVnDYDJJwSI/N/l4KrYu8Fmq1HDqrG2mlXDj09sDQq4NF6aXxsxemVYyb4asv2d2t60aW0Dsejih1VYqDvKYt8j9fab1/rHfe7V6Z1BGqYoW+VkL+3ao9b+/1skANqXWsBq9iGwNoYyHeuljboOwYPGU7ubkh4+SxH5SjfcCkhVYGFCKUVZDT1nklBei9ynLwmNlcgAAVJ1J0BPlq6Z/XpeDhm5OIN0cgq7cDrs3bgWaXUqyib0gppT4Vt1BOZph0Yr7aguTo8PAepmgqMeY5PNjcUEqce8Nx4kFT+N8k/aovkCAJOuCgLasw8NZNTQQDys6GYCaECC3NWaU1UBct8wZ2oHd25AA=) See BulkLoopOld vs. BulkLoopNew

According to the agner list of instruction latencies add with memory operand has a latency which is several times higher than the register version, I guess that explains the performance difference.

Background:
@samsosa opened a [PR](https://github.com/Tornhoof/MetroHash/pull/2) in MetroHash (a rather old project where I experimented with Span etc., it's a port of a non-cryptographic hash algorithm) and noted that assigning a local variable to a byref argument increases the performance of the bulk loop, the bulk loop is the main processing loop.

Example Code for repro, including benchmark, this is basically a stripped version of metrohash's bulk loop, this happens with both the latest 5.0 nightly as with 3.1

I couldn't find an existing issue regarding this particular CQ problem.

``` ini

BenchmarkDotNet=v0.12.1, OS=Windows 10.0.19041.329 (2004/?/20H1)
Intel Core i9-9900K CPU 3.60GHz (Coffee Lake), 1 CPU, 16 logical and 8 physical cores
.NET Core SDK=5.0.100-preview.8.20363.2
  [Host]     : .NET Core 5.0.0 (CoreCLR 5.0.20.36102, CoreFX 5.0.20.36102), X64 RyuJIT
  DefaultJob : .NET Core 5.0.0 (CoreCLR 5.0.20.36102, CoreFX 5.0.20.36102), X64 RyuJIT


```
|           Method |      Mean |    Error |   StdDev |
|----------------- |----------:|---------:|---------:|
|    WithLocalCopy |  83.66 ns | 0.796 ns | 0.744 ns |
| WithoutLocalCopy | 189.30 ns | 1.714 ns | 1.519 ns |

<details>

```csharp

using System.Runtime.CompilerServices;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;

namespace RegisterAllocRepro
{
    class Program
    {
        static void Main(string[] args)
        {
            BenchmarkRunner.Run<Repro>();
        }
    }

    public class Repro
    {
        private const ulong K0 = 0xC83A91E1;
        private const ulong K1 = 0x8648DBDB;
        private const ulong K2 = 0x7BDEC03B;
        private const ulong K3 = 0x2F5870A5;

        private static readonly byte[] _Data = new byte[1024];

        [Benchmark]
        public void WithLocalCopy()
        {
            ulong firstState = 0;
            ulong secondState = 0;
            ulong thirdState = 0;
            ulong fourthState = 0;
            int offset = 0;
            int count = 1024;
            var input = _Data;
            ref var start = ref input[0];
            BulkLoopNew(ref firstState, ref secondState, ref thirdState, ref fourthState, ref start, ref offset,
                count);
            FinalizeHash(ref firstState, ref secondState, ref start, ref offset, count);
        }

        [Benchmark]
        public void WithoutLocalCopy()
        {
            ulong firstState = 0;
            ulong secondState = 0;
            ulong thirdState = 0;
            ulong fourthState = 0;
            int offset = 0;
            int count = 1024;
            var input = _Data;
            ref var start = ref input[0];
            BulkLoopOld(ref firstState, ref secondState, ref thirdState, ref fourthState, ref start, ref offset,
                count);
            FinalizeHash(ref firstState, ref secondState, ref start, ref offset, count);
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        private static void BulkLoopOld(ref ulong firstState, ref ulong secondState, ref ulong thirdState,
            ref ulong fourthState, ref byte b, ref int offset, int count)
        {
            while (offset <= count - 32)
            {
                firstState += Cast<ulong>(ref b, offset) * K0;
                offset += 8; // add dword ptr [rax], 8
                firstState = RotateRight(firstState, 29) + thirdState;
                secondState += Cast<ulong>(ref b, offset) * K1;
                offset += 8;// add dword ptr [rax], 8
                secondState = RotateRight(secondState, 29) + fourthState;
                thirdState += Cast<ulong>(ref b, offset) * K2;
                offset += 8;// add dword ptr [rax], 8
                thirdState = RotateRight(thirdState, 29) + firstState;
                fourthState += Cast<ulong>(ref b, offset) * K3;
                offset += 8;// add dword ptr [rax], 8
                fourthState = RotateRight(fourthState, 29) + secondState;
            }
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        private static void BulkLoopNew(ref ulong firstState, ref ulong secondState, ref ulong thirdState,
            ref ulong fourthState, ref byte b, ref int finalOffset, int count)
        {
            var offset = finalOffset;
            while (offset <= count - 32)
            {
                firstState += Cast<ulong>(ref b, offset) * K0;
                offset += 8; // add r11d, 8
                firstState = RotateRight(firstState, 29) + thirdState;
                secondState += Cast<ulong>(ref b, offset) * K1;
                offset += 8; // add r11d, 8
                secondState = RotateRight(secondState, 29) + fourthState;
                thirdState += Cast<ulong>(ref b, offset) * K2;
                offset += 8; // add r11d, 8
                thirdState = RotateRight(thirdState, 29) + firstState;
                fourthState += Cast<ulong>(ref b, offset) * K3;
                offset += 8; // add r11d, 8
                fourthState = RotateRight(fourthState, 29) + secondState;
            }

            finalOffset = offset;
        }

        private static void FinalizeHash(ref ulong firstState, ref ulong secondState, ref byte b, ref int offset,
    int count)
        {
            var end = offset + (count & 31);

            if (end - offset >= 16)
            {
                firstState += Cast<ulong>(ref b, offset) * K2;
                offset += 8;
                firstState = RotateRight(firstState, 33) * K3;
                secondState += Cast<ulong>(ref b, offset) * K2;
                offset += 8;
                secondState = RotateRight(secondState, 33) * K3;
                firstState ^= RotateRight(firstState * K2 + secondState, 45) * K1;
                secondState ^= RotateRight(secondState * K3 + firstState, 45) * K0;
            }

            if (end - offset >= 8)
            {
                firstState += Cast<ulong>(ref b, offset) * K2;
                offset += 8;
                firstState = RotateRight(firstState, 33) * K3;
                firstState ^= RotateRight(firstState * K2 + secondState, 27) * K1;
            }

            if (end - offset >= 4)
            {
                secondState += Cast<uint>(ref b, offset) * K2;
                offset += 4;
                secondState = RotateRight(secondState, 33) * K3;
                secondState ^= RotateRight(secondState * K3 + firstState, 46) * K0;
            }

            if (end - offset >= 2)
            {
                firstState += Cast<ushort>(ref b, offset) * K2;
                offset += 2;
                firstState = RotateRight(firstState, 33) * K3;
                firstState ^= RotateRight(firstState * K2 + secondState, 22) * K1;
            }

            if (end - offset >= 1)
            {
                secondState += Unsafe.Add(ref b, offset) * K2;
                secondState = RotateRight(secondState, 33) * K3;
                secondState ^= RotateRight(secondState * K3 + firstState, 58) * K0;
            }

            firstState += RotateRight(firstState * K0 + secondState, 13);
            secondState += RotateRight(secondState * K1 + firstState, 37);
            firstState += RotateRight(firstState * K2 + secondState, 13);
            secondState += RotateRight(secondState * K3 + firstState, 37);
        }


        [MethodImpl((MethodImplOptions.AggressiveInlining))]
        private static T Cast<T>(ref byte b, int offset)
        {
            return Unsafe.As<byte, T>(ref Unsafe.Add(ref b, offset));
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        private static ulong RotateRight(ulong x, int r)
        {
            return (x >> r) | (x << (64 - r));
        }
    }
}

```
</details>

category:cq
theme:register-allocator
skill-level:expert
cost:large

**Comment by AndyAyersMS** [comment]
Thanks for the report. Was half expecting this to be about the implicit byref promotion heuristic, but it looks like it's more likely that the jit is not able to disambiguate the dereferences despite knowing the actual variables involved, so perhaps some form of forward sub.

Am going to mark this as future. We still need to take a closer look to figure out exactly what is going on.
